source/extensions/quic_listeners/quiche/platform/quic_reference_counted_impl.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include <memory>
+
+#include "quiche/quic/platform/api/quic_export.h"
+
+namespace quic {
+
+/** The implementation of reference counted object is merely wrapping
+ * std::shared_ptr. So QuicReferenceCountedImpl class does not do anything
+ * related to reference counting as shared_ptr already takes care of that. But
+ * it customizes destruction to provide a interface for shared_ptr to destroy
+ * the object, because according to the API declared in QuicReferenceCounted,
+ * this class has to hide its destructor.
+ */
+class QuicReferenceCountedImpl {
+public:
+  QuicReferenceCountedImpl() {}
+
+  // Expose destructor through this method.
+  static void destroy(QuicReferenceCountedImpl* impl) { delete impl; }
+
+protected:
+  // Non-public destructor to match API declared in QuicReferenceCounted.
+  virtual ~QuicReferenceCountedImpl() {}
+};
+
+template <typename T> class QuicReferenceCountedPointerImpl {
+public:
+  static_assert(std::is_base_of<QuicReferenceCountedImpl, T>::value,
```

Comment:
Change to std::is_base_of<QuicReferenceCounted, T>::value?

Reply:
This will cause a circular dependency because QuicReferenceCounted inherits from QuicReferenceCountedImpl.